### PR TITLE
Allow split compilation for native compute files

### PR DESCRIPTION
**Pull request by ppedrot**
~~This is a experiment for people to discuss about.~~ This is now deemed ready.

See https://github.com/coq/ceps/pull/48 for the bigger picture.

1. This PR introduces a new `coqnative` program that converts `vo` files to their `cmxs` native library.
2. The hybrid Makefile build was adapted: if `-native-compiler` **configure** option is set to `yes`, it will shortcut the native library generation of `coqc` by passing to it the `-native-compiler ondemand` **runtime** option and call instead `coqnative` on the generated `vo` file
3. ~~`coq_makefile` does the same basing its decision on `COQMF_COQ_NATIVE_COMPILER_DEFAULT`, which is set to the configure value by default.~~ This is deferred to a subsequent PR.

- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

### CODE REVIEW

**Comment by ejgallego** [comment]
That's nice; we could indeed reorganize code quite a lot IMO if we decide to go this path.

**Comment by ppedrot** [comment]
The PR in its current state compiles the standard library with Makefile. I need to adapt both coq_makefile and the dune build to test this more properly. For dune I have little idea of what to do, can you help @ejgallego?

**Comment by ejgallego** [comment]
> For dune I have little idea of what to do, can you help @ejgallego?

I can take care of that, but what's the problem exactly? If that's to compile coqnative it is trivial, just add it in the proper place to /topbin/dune

**Comment by ppedrot** [comment]
@ejgallego wasn't the PR adding native compilation to the dune build merged already? If not then indeed there is little more to do for now.

**Comment by ppedrot** [comment]
I rebased after the merge of #13352. See the top post for up-to-date status.

File: tools/CoqMakefile.in
Comment (gares) [inline tools/CoqMakefile.in:728]: @ppedrot can't you replace this test with a test on the .vo ? (implicit or not in the call to COQNATIVE)

Comment (ppedrot) [reply tools/CoqMakefile.in:728]: What should be tested exactly? vo files do not contain anything native-related.

Comment (ppedrot) [reply tools/CoqMakefile.in:728]: (This is the whole point btw: vo files should be native-agnostic if we ever want to have a split package infrastructure.)

Comment (gares) [reply tools/CoqMakefile.in:728]: But you said that calling COQNATIVE on a vo compiled with -native off raises an error. How is it.

Comment (ppedrot) [reply tools/CoqMakefile.in:728]: That's not exactly what I said. If a vo file has dependencies whose native files are missing, then the native compilation process will fail. This is unrelated to coqnative per se, coqc behaves exactly the same.

As I argued in the CEP, `-native-compiler no` means two things at once that do not need to be correlated in general, namely 1. use `native_compute` and 2. generate cmxs files when compiling vo. coqnative is insensitive to the configure flag insofar as if it has accesses to the vo, it doesn't care about the origin of the cmxs files (either from coqc or coqnative). Furthermore, it could very well generate recursively the missing dependencies from the vo files.

**Comment by ejgallego** [comment]
Nice, note that once we do the refactoring discussed in https://github.com/coq/coq/pull/13418 should make the duplicated code go away. Addressing it is pretty high in my todo [as it will also remove a bunch of duplicate code in jscoq/serapi].

**Comment by ppedrot** [comment]
I don't know how much sharing is expected with `coqtop`. If anything, `coqnative` is more similar to `coqchk` than to `coqc`, because it only needs access to the kernel facing API and nothing else. In particular, most `coqc` options are meaningless for `coqnative`, and similarly to `coqchk` there is no difference between `-R` and `-Q`. I'd rather have a minilibrary for this kind of mechanisms instead of hooking inside vernac.

(By the way, the simple compiler does wonders for memory consumption on perennial compilation, or said otherwise, when the hell do we get rid of the STM for `-batch` mode? This will be my top priority for the CUDW.)

**Comment by ejgallego** [comment]
> I'd rather have a minilibrary for this kind of mechanisms instead of hooking inside vernac.

This is what we proposed on the discussion of the PR, the amount of code duplicated is not large but indeed it has shown to create problems, for example due to safe_typing flags. IMHO it'd be worth to try to refactor a bit.

Note that you already hook on vernac due to `Library` but indeed this can be easily refined [for CUDW we can remove the makefile part of the build system for ML and then having more granular libs is easy, I found a reasonable way to hook make to dune]

> the simple compiler does wonders for memory consumption on perennial compilation, or said otherwise, when the hell do we get rid of the STM for `-batch` mode? This will be my top priority for the CUDW.

That'd be nice, as indeed the simple compiler wants to use the API in the same way as coq-lsp / flèche, we need to tweak the vernac API a bit still, in particular so we enforce correct treatment of side-effects w.r.t. envs [which can eat a bit of memory if choice is not available in the upper layers]

**Comment by ppedrot** [comment]
> Note that you already hook on vernac due to Library but indeed this can be easily refined

Nope, it's a local module that doesn't rely on the Library module. `coqnative` uses the very same code as `coqchk`, and the latter doesn't even link Library.

**Comment by ejgallego** [comment]
> Nope, it's a local module that doesn't rely on the Library module. `coqnative` uses the very same code as `coqchk`, and the latter doesn't even link Library.

We should maybe refactor these into a lib? That should be easy to do, hopefully next week I can push a couple of PRs regading common lib refactoring.

Other than that this should be OK, and I can be the assignee. I will teach dune about `coqnative` so things should work out of the box.

Note that the unsoundness in the make-based build system is a bit exacerbated by this PR, in the sense that for example if you remove `cmxs` files, they won't get rebuild properly, but that's not exclusive to this PR.

**Comment by ejgallego** [comment]
I'd say this looks pretty good, I'd be happy to assign if you want. Main TODO is to profit from the recent work by @gares doing the sysinit stuff as to remove the small bits of duplication in the cmdline tool, I could take care of it if you want as I've got some experience [and I've written myself quite a few similar tools in serapi [sercomp / sername, etc...]

**Comment by ppedrot** [comment]
@ejgallego do not hesitate to push patches atop of this PR, I don't have a lot of time right now to work on it and it'd nice to get at least the binary shipped with 8.14, regardless of the build system / package management issues.

**Comment by ejgallego** [comment]
@ppedrot this looks good to me, I'd rather have this merged as is [once we fix the test suite], and then do the refactoring that will move the loadpath handling to a common lib on a different PR, it will be easier to have the tool merged, it will also allow us to release dune 2.9 that has rules for `.vo -> .cmxs` processing.

**Comment by ejgallego** [comment]
Failures looks spurious tho there are a couple that deal with too large build times / artifacts so we should bench this.

**Comment by ppedrot** [comment]
The bench won't run native compilation though, should we hack it?

**Comment by ejgallego** [comment]
> The bench won't run native compilation though, should we hack it?

As you want, not sure it is worth it.

I've pushed an ultra minor fix to get the CI going.

**Comment by ejgallego** [comment]
> I've pushed an ultra minor fix to get the CI going

CI seems OK, except for the missing hackery in configure for the OSX something.

File: topbin/coqnative_bin.ml
Comment (gares) [inline topbin/coqnative_bin.ml:267]: Is there a good reason not to use sysinit?

Comment (ppedrot) [reply topbin/coqnative_bin.ml:267]: Same reason the checker doesn't use it: it's too high in the dependency chain.

Comment (gares) [reply topbin/coqnative_bin.ml:267]: The checker has the, to me good, reason to not depend on the higher layers.
I don't see how this very same reason applies here, but maybe there are others.

Comment (gares) [reply topbin/coqnative_bin.ml:267]: More precisely, these is some code about parsing and interpreting -R, -Q, -I here which has no reason to be duplicate.

Comment (ppedrot) [reply topbin/coqnative_bin.ml:267]: It has the same semantics as the checker, i.e. `-R / -Q` are identified since relative paths are not used. I believe @ejgallego 's work is precisely going to extract the relevant parts from coqinit to put them at the lowest point possible so that all Coq binaries can use them as a library. Otherwise your complaint also applies to coqchk and coqdep.

Also, the vast majority of flags accepted in coqinit are totally irrelevant for that tool.

Comment (gares) [reply topbin/coqnative_bin.ml:267]: Fair enough, but then maybe we should this PR depend on that one. (It is a new command and is still a draft)
@ejgallego what is the status of the split of sysinit ?

Comment (ppedrot) [reply topbin/coqnative_bin.ml:267]: See https://github.com/coq/coq/pull/13287#issuecomment-808747055

Comment (gares) [reply topbin/coqnative_bin.ml:267]: So you guys think it is a good idea to make a dune release which calls a tool merged in coq master ten minutes before?

I'm find merging as is and clean up later, but doing it for that reason is ********

Comment (ppedrot) [reply topbin/coqnative_bin.ml:267]: I personally don't care about the dune release, I just want the coqnative binary to be shipped in the next Coq release. Even without support from any build system it would help some users who repeatedly asked for features that would be solved with coqnative.

**Comment by SkySkimmer** [comment]
No topbin/dune change?

File: topbin/coqnative_bin.ml
Comment (ejgallego) [reply topbin/coqnative_bin.ml:267]: @gares actually there are two separate things here:
- unifying coqchk / coqdep / coqnative code: I have a PR doing that using a sites/findlib inspired API which IMHO is quite Cool as avoid misuses by making paths opaque, but that depends on #13617 [for bootstrap reasons, this refactoring work is super hard to with out make-based OCaml build system]
- having dune support coqnative, this is largely orthogonal and not a problem, as dune can check if coqnative is present in the workspace and enable the rules conditionally on that.

Comment (ejgallego) [reply topbin/coqnative_bin.ml:267]: My point of view as whether to merge is that any delay to a PR usually has an exponential cost, it is way more easier for everyone to merge as soon a possible.

Comment (ejgallego) [reply topbin/coqnative_bin.ml:267]: For this PR I am very happy to have it soon in master and we can tweak it, than having a complex layer of inter-depending PRs

Comment (ejgallego) [reply topbin/coqnative_bin.ml:267]: > @ejgallego what is the status of the split of sysinit 

@gares I think I am mostly done with the "use dune to build OCaml" so will submit a draft PR pretty soon; I'd personally would merge whichever is ready first, this tool will indeed use the same code than coqchk.

**Comment by JasonGross** [comment]
Something about this seems to blow up the build time of fiat-crypto legacy.  It would be nice to know what's going on and how to fix it.  (My current guess is that the CI is compiling to native code all the .vo files that it previously did not, and these .vo files are filled with enormous PHOASTs, and the fix is just to get fcl to pass the ondemand flag instead?)

**Comment by ppedrot** [comment]
@JasonGross I can't really tell since I do not understand the fcl build system. This PR is just moving the compilation of the native files to an external process, so you need to be sure not to call coqnative on the fcl files. But once again I have no idea what needs to be done for this not to happen.

File: tools/CoqMakefile.in
Comment (JasonGross) [inline tools/CoqMakefile.in:731]: I think it would be nice to allow timing for these files.  Maybe add something like
```make
TIMECMD_NATIVE?=$(TIMECMD)
# Use command time on linux, gtime on Mac OS
TIMEFMT_NATIVE?="$*.cmxs (real: %e, user: %U, sys: %S, mem: %M ko)"
ifneq (,$(TIMED))
ifeq (0,$(shell command time -f "" true >/dev/null 2>/dev/null; echo $$?))
STDTIME_NATIVE?=command time -f $(TIMEFMT_NATIVE)
else
ifeq (0,$(shell gtime -f "" true >/dev/null 2>/dev/null; echo $$?))
STDTIME_NATIVE?=gtime -f $(TIMEFMT_NATIVE)
else
STDTIME_NATIVE?=command time
endif
endif
else
STDTIME_NATIVE?=command time -f $(TIMEFMT_NATIVE)
endif
```
and then add
```make
TIMER_NATIVE=$(if $(TIMED), $(STDTIME_NATIVE), $(TIMECMD_NATIVE))
```
and then use
```suggestion
	$(HIDE)$(TIMER_NATIVE) $(COQNATIVE) $(COQLIBS) $*.vo
```

**Comment by JasonGross** [comment]
@ppedrot The issue is that `export COQEXTRAFLAGS='-native-compiler no'` is apparently no longer enough to suppress the compilation of .v files to native code, and yet the ci script has not changed at
https://github.com/coq/coq/blob/f7e05393addb9a45fa1eaef2a72da9b2c4c14396/dev/ci/ci-fiat_crypto_legacy.sh#L12

The build system of fcl does use coq_makefile-made makefiles, so fixing this should not be specific to the build system of fcl.

**Comment by JasonGross** [comment]
Btw, I expect that you will also need an overlay for HoTT, since we rely on `-arg -native-compiler -arg ondemand` (or `-arg no`) in `_CoqProject` to suppress building of the native compiler libraries, and it looks like this PR will make that not work anymore?

**Comment by ppedrot** [comment]
I've just performed a botched rebase after the merge of the dune-make hybrid but it doesn't work. For some reason the coqnative binary cannot be built due to a missing rule but I don't understand who is responsible for that.

**Comment by SkySkimmer** [comment]
topbin/dune should have something about coqnative

**Comment by ppedrot** [comment]
Let's see if this works.

**Comment by ppedrot** [comment]
A bit of progress has been made. Now the binary coqnative compiles correctly and is invoked on the right files, but for some reason the generated files are nowhere to be found. I believe that I need to perform some magic to move them where they were compiled like vo files.

**Comment by ppedrot** [comment]
OK, so the problem lies somewhere around `Nativelib.include_dirs`. The list of paths passed to the ocamlopt call is incorrect. I need to understand what is expected to happen.

**Comment by ejgallego** [comment]
> OK, so the problem lies somewhere around `Nativelib.include_dirs`. The list of paths passed to the ocamlopt call is incorrect. I need to understand what is expected to happen.

Let me know in Zulip if you need help, this is indeed tricky.

**Comment by ejgallego** [comment]
@ppedrot , would it make sense to split this PR in 3 parts:
- the coqnative part, which can be merged now [so I can test the dune rules and include coqnative in coq.boot refactoring]
- the build system changes, to use coqnative
- the cleanup of coqc so it doesn't output native anymore [that would have to target 8.15 I understand]

**Comment by ppedrot** [comment]
This should be working now.

@ejgallego as you wish, I can indeed split this PR in two parts. The `coqc` cleanup part is not very urgent since it is easy to deactivate native compilation on a per-command basis.

**Comment by ejgallego** [comment]
> This should be working now.

Then I propose you undraft and we merge in one piece; I proposed the split as to speed up the process, but if this is working I don't see a reason to do it.

**Comment by ppedrot** [comment]
OK, so I remember the source of the issue: in the new coq_makefile code, setting COQEXTRAFLAGS to unset deactivate compilation is not correct, but our CI infrastructure (and many external developments) rely on this flag for that. It's probably enough not to include the coq_makefile commit to make things work as before though. Let me check if it works as intended.

**Comment by ppedrot** [comment]
... and I can't test locally because I still don't know how to get a working coq_makefile in hybrid mode. `make world && make ci-unimath` is not enough. Sigh.

**Comment by ejgallego** [comment]
> ... and I can't test locally because I still don't know how to get a working coq_makefile in hybrid mode. `make world && make ci-unimath` is not enough. Sigh.

I'm testing this, one sec.

**Comment by ejgallego** [comment]
> I'm testing this, one sec.

Bah it seems I exactly forgot to add the dependency to the CoqMakefile.in ...

**Comment by ejgallego** [comment]
> Bah it seems I exactly forgot to add the dependency to the CoqMakefile.in ...

Was more subtle, if you call `_build_vo/default/bin/coq_makefile` things work, it is the script in ci common that sets something wrong it seems.

**Comment by ppedrot** [comment]
Reverting the coq_makefile commit seems to make things work smoothly. I'll remove it from this PR and submit it in a separate one, so that we can merge this one quickly.

**Comment by ejgallego** [comment]
> Reverting the coq_makefile commit seems to make things work smoothly. I'll remove it from this PR and submit it in a separate one, so that we can merge this one quickly.

For the ci- problem see #14225 , indeed ci-common went broken in master after recent changes.

**Comment by ejgallego** [comment]
You can tweak `ci-common` locally as in that PR and you should be able to proceed with `make ci-unimath`, thanks for your patiente.

**Comment by ppedrot** [comment]
So, even with #14225, compiling external developments
- in the current state of this PR (i.e. without touching to the coq_makefile script)
- with native enabled
- a local build

fails due to `coqc` not being able to find the native compilation files in its path, namely the `Nativevalues` file from the kernel. It's not a problem of `coqnative` so it's unclear whether this PR has anything to do with it, although probably not.

Otherwise, the CI passes and correctly generates the native files from what I gather (and doesn't when COQEXTRAFLAG is set), so it means that after an install it works as expected. Since we have other issues with hybrid + local, I am going to mark this PR as ready, it can't really make things worse for this setting.

**Comment by ejgallego** [comment]
> fails due to `coqc` not being able to find the native compilation files in its path, namely the `Nativevalues` file from the kernel. It's not a problem of `coqnative` so it's unclear whether this PR has anything to do with it, although probably not.

Ok, thanks for the feedback, let me reproduce and fix.

**Comment by ejgallego** [comment]
> Ok, thanks for the feedback, let me reproduce and fix.

Fix pushed to the fix PR.

**Comment by ppedrot** [comment]
Added doc and changelog.

**Comment by ejgallego** [comment]
Nice, thanks @ppedrot ; I'd like to merge soon, but actually two of the most important reviewers for this PR may be away; let's cc them and see what happens next week (cc: @maximedenes @silene )

**Comment by Zimmi48** [comment]
Even if this PR is very related to native-compute, they are not actually the code owners for the files that are changed...

File: doc/changelog/08-cli-tools/13287-native-vo-to-cmxs-compiler.rst
Comment (Zimmi48) [inline doc/changelog/08-cli-tools/13287-native-vo-to-cmxs-compiler.rst:2]: No need to repeat "Added":
```suggestion
  A standalone `coqnative` binary that performs native compilation
```

File: doc/changelog/08-cli-tools/13287-native-vo-to-cmxs-compiler.rst
Comment (Zimmi48) [inline doc/changelog/08-cli-tools/13287-native-vo-to-cmxs-compiler.rst:5]: I'm not sure if mentioning "the hybrid build system" is actually necessary.

```suggestion
  native compilation. The hybrid build system was adapted to perform a split
  compilation on the stdlib
```

File: doc/changelog/08-cli-tools/13287-native-vo-to-cmxs-compiler.rst
Comment (Zimmi48) [inline doc/changelog/08-cli-tools/13287-native-vo-to-cmxs-compiler.rst:2]: Probably worth adding a reference to the documentation of `coqnative`.

File: doc/sphinx/practical-tools/utilities.rst
Comment (Zimmi48) [inline doc/sphinx/practical-tools/utilities.rst:686]: I don't recall `-nI` being documented anywhere (I was not even aware of its existence until recently). Anyway, if this one too is a no-op when passed to `coqnative`, that should probably be made explicit.

**Review by Zimmi48 — APPROVED** [review]

File: doc/sphinx/practical-tools/utilities.rst
Comment (ejgallego) [reply doc/sphinx/practical-tools/utilities.rst:686]: `-nI` is really internal so that is the reason it was not documented.

File: doc/sphinx/practical-tools/utilities.rst
Comment (ejgallego) [inline doc/sphinx/practical-tools/utilities.rst:693]: I'd suggest adding a warning when -I is used.

**Review by ejgallego — APPROVED** [review]
Looks pretty good as a first step!

We need to factor out some code, in particular for sysinit and also coq.boot, but the duplication is a reality in the tree today so I think it is best to handle those in the corresponding PRs dealing with moving these bits lower in the dep chain.

File: doc/sphinx/practical-tools/utilities.rst
Comment (ppedrot) [reply doc/sphinx/practical-tools/utilities.rst:686]: Should I remove it from the doc?

File: doc/sphinx/practical-tools/utilities.rst
Comment (ppedrot) [reply doc/sphinx/practical-tools/utilities.rst:693]: The reason I have included it is that coq_makefile stores the `-R / -Q` and `-I` flags in the same variable. If we start spamming warnings for any library that includes ML plugins it's not going to be great, and fixing coq_makefile not to do that requires changes that may break developments out there that rely on the variable in their own scripts.

Comment (ejgallego) [reply doc/sphinx/practical-tools/utilities.rst:693]: Ok, let's address this separately then

File: doc/sphinx/practical-tools/utilities.rst
Comment (ejgallego) [reply doc/sphinx/practical-tools/utilities.rst:686]: Seems a good idea, as it is likely we can get rid of this option for 8.14

**Comment by ejgallego** [comment]
I will tentatively try to merge Friday.

**Comment by ejgallego** [comment]
@ppedrot will need a rebase after the PR fixing the windows build is merged, "just in case"

**Comment by ppedrot** [comment]
OK, ping me when this is done.

**Comment by ejgallego** [comment]
> OK, ping me when this is done.

Just merged the Windows CI in master.

**Comment by ejgallego** [comment]
Green tick :) Here we go!

@coqbot: merge now

**Comment by ejgallego** [comment]
@ppedrot there is a bug here that we didn't catch, see the CI log for this PR merge:
https://gitlab.com/coq/coq/-/jobs/1239611568
```
install: cannot stat '_build_vo/default//lib/coq/user-contrib/Ltac2/.coq-native/NLtac2_Ltac1.o': No such file or directory
install: cannot stat '_build_vo/default//lib/coq/user-contrib/Ltac2/.coq-native/NLtac2_Printf.o': No such file or directory
...
```
the log for the previous commit was OK: https://gitlab.com/coq/coq/-/jobs/1240127653

**Comment by ppedrot** [comment]
Ltac2 is the only plugin in the `user-contrib` directory, is the build system doing something fancy with it?

**Comment by ejgallego** [comment]
Nothing too fancy, it is just and additional list of directories that is processed.

**Comment by ppedrot** [comment]
Something wrong happens when compiling the Ltac2 files:
```
COQNATIVE user-contrib/Ltac2/Env.vo
Error: File Init.vo not found in loadpath
```

**Comment by ppedrot** [comment]
This looks like COQLIB is not correctly set, again.

**Comment by ejgallego** [comment]
> This looks like COQLIB is not correctly set, again.

It doesn't look like it, but more like a missing build dep, as this is `Init.v` in Ltac2

**Comment by ppedrot** [comment]
No, the vo file is in the `build_vo` directory in the right place. The coqlib option passed to coqnative seems wrong though, this is a command line that fails.
```
pm@ouranos:/tmp/coq$ _build/install/default/bin/coqnative -nI _build/default/kernel/.kernel.objs/byte -coqlib . -R _build_vo/default//lib/coq/theories Coq _build_vo/default//lib/coq/user-contrib/Ltac2/Ident.vo
```

Note the `-coqlib .`.

**Pull request merged**
